massa-final-state/src/controller_trait.rs

```diff
@@ -0,0 +1,93 @@
+use massa_async_pool::AsyncPool;
+use massa_db_exports::{DBBatch, ShareableMassaDBController};
+use massa_executed_ops::ExecutedDenunciations;
+use massa_hash::Hash;
+use massa_ledger_exports::LedgerController;
+use massa_models::{operation::OperationId, slot::Slot};
+use massa_pos_exports::PoSFinalState;
+use massa_versioning::versioning::MipStore;
+
+use crate::{FinalStateError, StateChanges};
+
+/// Trait for final state controller.
+#[cfg_attr(feature = "test-exports", mockall::automock)]
+pub trait FinalStateController: Send + Sync {
+    /// Get the fingerprint (hash) of the final state.
+    /// Note that only one atomic write per final slot occurs, so this can be safely queried at any time.
+    fn get_fingerprint(&self) -> Hash;
+
+    /// Get the slot at the end of which the final state is attached
+    fn get_slot(&self) -> Slot;
+
+    /// Gets the hash of the execution trail
+    fn get_execution_trail_hash(&self) -> Hash;
+
+    /// Reset the final state to the initial state.
+    ///
+    /// USED ONLY FOR BOOTSTRAP
+    fn reset(&mut self);
+
+    /// Performs the initial draws.
+    fn compute_initial_draws(&mut self) -> Result<(), FinalStateError>;
+
+    /// Applies changes to the execution state at a given slot, and settles that slot forever.
+    /// Once this is called, the state is attached at the output of the provided slot.
+    ///
+    /// Panics if the new slot is not the one coming just after the current one.
+    fn finalize(&mut self, slot: Slot, changes: StateChanges);
+
+    /// After bootstrap or load from disk, recompute all the caches.
+    fn recompute_caches(&mut self);
+
+    /// Deserialize the entire DB and check the data. Useful to check after bootstrap.
+    fn is_db_valid(&self) -> bool;
+
+    /// Initialize the execution trail hash to zero.
+    fn init_execution_trail_hash_to_batch(&mut self, batch: &mut DBBatch);
+
+    /// Get ledger
+    #[allow(clippy::borrowed_box)]
+    fn get_ledger(&self) -> &Box<dyn LedgerController>;
+
+    /// Get ledger mut
+    fn get_ledger_mut(&mut self) -> &mut Box<dyn LedgerController>;
+
+    /// Get async pool
+    fn get_async_pool(&self) -> &AsyncPool;
+
+    /// Get pos state
+    fn get_pos_state(&self) -> &PoSFinalState;
+
+    /// Get pos state mut
+    fn get_pos_state_mut(&mut self) -> &mut PoSFinalState;
+
+    /// check if an operation is in the executed ops
+    fn executed_ops_contains(&self, op_id: &OperationId) -> bool;
+
+    /// Get the executed status ops
+    fn get_ops_exec_status(&self, batch: &[OperationId]) -> Vec<Option<bool>>;
```

Comment:
Why do you leak executed_ops related functions here? Why can't you do the same getter as with AsyncPool or PosFinalState ?

Reply:
Because I believe how Async Pool and PoS is done it's badly done you shouldn't have a &ConcreteType in a trait it's a short term solution. The FInalState should just expose his methods and if necessary call another module (or make the module out of the final state). I made it correctly for the executed because there was only two methods and to test it